### PR TITLE
Document the deprecation policy and addd the deprecate_parameter decorator to deprecate parameters

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

This is a draft implementation to deprecate a parameter but keep backward compatibility.

For example, the `Figure.plot()` function has the `sizes` parameter, but we want to rename it to `size`, and keep backward compatibility. We can add the `deprecate_parameter` decorator to the `Figure.plot()` method:

```
@deprecate_parameter("sizes", "size", "v0.4.0", "v0.6.0")
def plot():
   ...
```
It means the old parameter name `sizes` is deprecated since v0.4.0, and will be removed in v0.6.0.
The new parameter name is `size`.

When users use `sizes`, a warning will be raised.

References:

- https://stackoverflow.com/questions/2536307
- https://docs.python.org/3/library/warnings.html
- https://www.python.org/dev/peps/pep-0565/

Related to #1120.

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/helpers/decorators.py
Comment (weiji14) [inline pygmt/helpers/decorators.py:523]:
```suggestion
                    f"The '{oldname}' parameter is deprecated since {deprecate_version} and will"
```

File: pygmt/helpers/decorators.py
Comment (weiji14) [inline pygmt/helpers/decorators.py:524]:
```suggestion
                    f" be removed in {remove_version}; please use '{newname}' instead."
```

File: pygmt/src/plot.py
Comment (weiji14) [inline pygmt/src/plot.py:47]: I would use the keyword name here, make it easier to regex match when the time comes to deprecate something. Also, do we want to deprecate this at v0.6.0 or v0.5.0?

```suggestion
@deprecate_parameter("sizes", "size", "v0.4.0", remove_version="v0.6.0")
```

File: pygmt/helpers/decorators.py
Comment (weiji14) [inline pygmt/helpers/decorators.py:528]: One more thing, I don't quite get what the `warnings.simplefilter` is here for? I see it's copied from the stackoverflow answer, but I'm still wondering why we can't just do `warnings.warn`.

File: pygmt/src/plot.py
Comment (seisman) [reply pygmt/src/plot.py]: > Also, do we want to deprecate this at v0.6.0 or v0.5.0?

As `sizes` is heavily used, I'd like to give users more time to change the old names to new names.

File: pygmt/helpers/decorators.py
Comment (seisman) [reply pygmt/helpers/decorators.py:528]: I haven't read the full documentation of `warnings`, so I can't answer the question now.

Comment (seisman) [reply pygmt/helpers/decorators.py:528]: > One more thing, I don't quite get what the `warnings.simplefilter` is here for?

The `DeprecationWarning` is ignored by default. So, if you remove the `warnings.simplefilter` calls and run the example below, you won't see the warnings. That's why `warnings.simplefilter` are necessary here.
```
import pygmt
fig = pygmt.Figure()
fig.basemap(region=[-5, 5, -5, 5], projection="X10c/10c", frame=True)
fig.plot(x=[-2, 2], y=[0, 0], style="c", sizes=[1, 2])
fig.plot(x=[-2, 2], y=[2, 2], style="c", size=[1, 2])
fig.savefig("symbols.pdf")
```

---

Here are some related descriptions about `DeprecationWarning` and `FutureWarning`. If I understand it correctly, we should use `FutureWarning` instead, right?

https://docs.python.org/3/library/warnings.html

> **DeprecationWarning**
> 
> Base category for warnings about deprecated features when those warnings are intended for other **Python developers** (ignored by default, unless triggered by code in `__main__`).
> 
> **FutureWarning**
> 
> Base category for warnings about deprecated features when those warnings are **intended for end users** of applications that are written in Python.
> 
> Changed in version 3.7: Previously DeprecationWarning and FutureWarning were distinguished based on whether a feature was being removed entirely or changing its behaviour. They are now distinguished based on their intended audience and the way they’re handled by the default warnings filters.

https://www.python.org/dev/peps/pep-0565/#id13

> DeprecationWarning: reported by default for code that runs directly in the __main__ module (as such code is considered relatively unlikely to have a dedicated test suite), but hidden by default for code in other modules. The intended audience is Python developers that are at risk of upgrades to their dependencies (including upgrades to Python itself) breaking their software (e.g. developers using Python to script environments where someone else is in control of the timing of dependency upgrades).
> 
> FutureWarning: reported by default for all code. The intended audience is users of applications written in Python, rather than other Python developers (e.g. warning about use of a deprecated setting in a configuration file format).
>

Comment (weiji14) [reply pygmt/helpers/decorators.py:506]: Yeah, looks like `FutureWarning` does look easier, but PEP 0565 states:

> For library and framework authors that want to ensure their API compatibility warnings are more reliably seen by their users, the recommendation is to use a custom warning class that derives from `DeprecationWarning` in Python 3.7+, and from `FutureWarning` in earlier versions.

So maybe we should stick to DeprecationWarning instead since we're Python 3.7+? I'm not so familiar with the history of DeprecationWarning/FutureWarning in Python to be honest, we should look at how other projects handle things e.g. `pandas`/`geopandas` perhaps.

Comment (seisman) [reply pygmt/helpers/decorators.py:506]: PEP 0565 also says:

> The standard library documentation will be updated to explicitly recommend the use of FutureWarning (rather than DeprecationWarning) for backwards compatibility warnings that are intended to be seen by users of an application.

Comment (seisman) [reply pygmt/helpers/decorators.py:506]: It seems both pandas and xarray use `FutureWarning` a lot:

- https://github.com/pandas-dev/pandas/blob/master/pandas/util/_depr_module.py
- https://github.com/pydata/xarray/blob/master/xarray/core/common.py

Comment (weiji14) [reply pygmt/helpers/decorators.py:506]: Ok, let's use FutureWarning then, would be good to remove the `warnings.simplefilter` lines too :smile:

File: pygmt/src/plot.py
Comment (seisman) [inline pygmt/src/plot.py:6]: FYI, I plan to revert the changes in `plot.py` before merging this PR, because the `deprecate_parameter` decorator is already tested in its docstrings, and the deprecation of `sizes` parameter need more changes and should be done in a separate PR.

File: pygmt/helpers/decorators.py
Comment (weiji14) [inline pygmt/helpers/decorators.py:447]: Maybe add an extra note here, feel free to edit my English.

```suggestion
    Decorator to deprecate a parameter.
    
    The old parameter name will be swapped automatically to the new parameter
    name, and users will receive a FutureWarning to inform them of the pending
    deprecation.
```

**Review by weiji14 — COMMENTED** [review]
Looks good, I would add a note about our deprecation policy (and how to use `@deprecate_parameter`) in the maintainer docs, maybe after #1185 is completed.

File: pygmt/helpers/decorators.py
Comment (seisman) [reply pygmt/helpers/decorators.py:447]: I also added one more note "Use this decorator below the ``use_alias`` decorator.", because `use_alias` will convert long-form parameters to short-form parameters.

**Comment by seisman** [comment]
> Looks good, I would add a note about our deprecation policy (and how to use `@deprecate_parameter`) in the maintainer docs, maybe after #1185 is completed.

#1185 is already merged. Do you want me to document the deprecation policy in this PR or in a separate PR?

**Comment by weiji14** [comment]
Document it here please.

File: doc/maintenance.md
Comment (meghanrjones) [inline doc/maintenance.md:159]:
```suggestion
- Incompatible changes should be documented in the release notes.
```

**Comment by seisman** [comment]
Ping @GenericMappingTools/pygmt-contributors for reviews and comments on this PR.

File: doc/maintenance.md
Comment (weiji14) [inline doc/maintenance.md:158]:
```suggestion
- Incompatible changes are only allowed in major and minor releases, not in
  patch releases.
```

File: doc/maintenance.md
Comment (weiji14) [inline doc/maintenance.md:168]: May want to wait for votes to come in on https://github.com/GenericMappingTools/pygmt/issues/1120#issuecomment-818377643 before committing.

```suggestion
- The `FutureWarning` warning should appear for 2-4 minor versions, depending on
```

File: doc/maintenance.md
Comment (weiji14) [inline doc/maintenance.md:174]:
```suggestion
To rename a function parameter, add the `@deprecated_parameter` decorator
before the function definition (but after the `@use_alias` decorator if it exists).
```

File: doc/maintenance.md
Comment (weiji14) [inline doc/maintenance.md:185]:
```suggestion

In this case, the old parameter name `sizes` is deprecated since v0.4.0, and will be
```

File: pygmt/helpers/decorators.py
Comment (weiji14) [inline pygmt/helpers/decorators.py:449]:
```suggestion
    The old parameter name will be automatically swapped to the new parameter
```

**Review by weiji14 — COMMENTED** [review]
Just some proofread edits. No outstanding issues on the `@deprecate_parameter` function.

File: pygmt/helpers/decorators.py
Comment (core-man) [inline pygmt/helpers/decorators.py:508]:
```suggestion
                    f" and will be removed in {remove_version}."
                    f" Please use '{newname}' instead."
```

File: pygmt/helpers/decorators.py
Comment (core-man) [inline pygmt/helpers/decorators.py:506]:
```suggestion
                    f"The '{oldname}' parameter has been deprecated since {deprecate_version}"
```

**Comment by core-man** [comment]
Woow~ Looks great :smile: Only some minor suggestions.

File: pygmt/helpers/decorators.py
Comment (core-man) [inline pygmt/helpers/decorators.py:502]: Add a test for this?

Comment (seisman) [reply pygmt/helpers/decorators.py:506]: Done in c73b80bdcb4f64bffd71a5eb1427ea9138954100.

**Review by core-man — APPROVED** [review]

**Review by weiji14 — APPROVED** [review]

**Pull request merged**
